### PR TITLE
Verify all branches are handled in conditional task in playbooks

### PR DESCRIPTION
**Pull request by DeanArbel**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/21850

## Description
Added is_condition_branches_handled_correctly to PlaybookValidator, which validates all condition branches are handled correctly.

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 1]
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
...
+                        try:
+                            if next_task_ids:
+                                task_condition_labels.remove(next_task_branch)
+                        except KeyError as e:
+                            print_error('Playbook has conditional task with unreachable next task: {}'.format(str(e)))
+                            self.is_valid = False
+                            return False
+                    # if there are task_condition_labels left then not all branches are handled
+                    if task_condition_labels:
+                        print_error('Playbook has conditional task with unhandled branches: {}'.format(
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/playbook.py]: use f-strings

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 2]
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
...
+                        label = condition.get('label')
+                        if label:
+                            task_condition_labels.add(label)
+                    next_tasks = task.get('nexttasks', {})
+                    for next_task_branch, next_task_ids in next_tasks.items():
+                        try:
+                            if next_task_ids:
+                                task_condition_labels.remove(next_task_branch)
+                        except KeyError as e:
+                            print_error('Playbook has conditional task with unreachable next task: {}'.format(str(e)))
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/playbook.py]: use f-strings
what will the error message look like? will it be any indicative?

Comment (DeanArbel) [reply demisto_sdk/commands/common/hook_validations/playbook.py]: Yeah, the str(e) of a key error is the key itself, so it will print the task ID.

**Comment by DeanArbel** [comment]
@Itay4 I removed the walrus operator, as it failed the build. demisto-sdk doesn't support py38, and according to @bakatzir, we are not planning on changing that soon.

File: demisto_sdk/commands/common/hook_validations/playbook.py (hunk 2 above)
Comment (Itay4) [reply demisto_sdk/commands/common/hook_validations/playbook.py]: please post screenshot of example

**Comment by Itay4** [comment]
> @Itay4 I removed the walrus operator, as it failed the build. demisto-sdk doesn't support py38, and according to @bakatzir, we are not planning on changing that soon.

of course we support py3.8... its just the build also runs py3.7
leave it for now, ill handle it later

**Comment by DeanArbel** [comment]
@Itay4 following your review I made the error messages more informative, please see the screenshot below as an example:
image

File: demisto_sdk/commands/common/hook_validations/playbook.py (hunk 2 above)
Comment (DeanArbel) [reply demisto_sdk/commands/common/hook_validations/playbook.py]: see comment below

**Comment by Itay4** [comment]
> @Itay4 following your review I made the error messages more informative, please see the screenshot below as an example:
> image

lets improve the english, something like `... has unhandled condition: <CONDITION-NAME>`
whats `#default` ?

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 3]
```diff
@@ -21,7 +23,8 @@ def is_valid_playbook(self, is_new_playbook=True):  # type: (bool) -> bool
                 self.is_valid_version(),
                 self.is_id_equals_name(),
                 self.is_no_rolename(),
-                self.is_root_connected_to_all_tasks()
+                self.is_root_connected_to_all_tasks(),
+                self.is_condition_branches_handled_correctly()
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/playbook.py]:
```suggestion
                self.is_all_condition_branches_handled()
```

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 4]
```diff
@@ -62,3 +64,36 @@ def is_no_rolename(self):  # type: () -> bool
             self.is_valid = False
             return False
         return True
+
+    def is_condition_branches_handled_correctly(self):  # type: () -> bool
```
[the remark this answers is not in the document]
Comment (Itay4) [reply demisto_sdk/commands/common/hook_validations/playbook.py]: add return type hint

File: demisto_sdk/commands/common/hook_validations/playbook.py [hunk 5]
```diff
@@ -65,6 +69,38 @@ def is_no_rolename(self):  # type: () -> bool
             return False
         return True
 
+    def is_condition_branches_handled_correctly(self):  # type: () -> bool
+        """Check whether the playbook conditional tasks has all optional branches handled
+
+        Return:
+            bool. if the Playbook handles all condition branches correctly.
+        """
+        is_unreachable: bool = True
```
Comment (Itay4) [inline demisto_sdk/commands/common/hook_validations/playbook.py]: the variable name here is misleading, lets improve it

File: demisto_sdk/commands/common/tests/playbook_test.py [hunk 6]
```diff
@@ -75,6 +121,18 @@ def test_is_added_required_fields(self, current_file, answer):
         (NEXT_TASKS_VALID_EXIST_2, True),
     ]
 
+    @pytest.mark.parametrize("current_file, answer", IS_NO_ROLENAME_INPUTS)
+    def test_is_added_required_fields(self, current_file, answer):
+        structure = mock_structure("", current_file)
+        validator = PlaybookValidator(structure)
+        assert validator.is_no_rolename() is answer
+
+    @pytest.mark.parametrize("current_file, answer", IS_CONDITIONAL_INPUTS)
```
Comment (Itay4) [inline demisto_sdk/commands/common/tests/playbook_test.py]: why do we name these `current_file` and `answer` ? this is confusing..

File: demisto_sdk/tests/test_files/playbook-Invalid_condition_1.yml [hunk 7]
```diff
@@ -0,0 +1,111 @@
+elasticcommonfields: {}
+id: Invalid_condition_1
```
Comment (Itay4) [inline demisto_sdk/tests/test_files/playbook-Invalid_condition_1.yml]: for both playbooks with invalid conditions, lets write why they're invalid

File: demisto_sdk/commands/common/tests/playbook_test.py (hunk 6 above)
Comment (DeanArbel) [reply demisto_sdk/commands/common/tests/playbook_test.py]: changed to `playbook_json` and `expected_result`

File: demisto_sdk/tests/test_files/playbook-Invalid_condition_1.yml (hunk 7 above)
Comment (DeanArbel) [reply demisto_sdk/tests/test_files/playbook-Invalid_condition_1.yml]: changed to `Invalid_condition_unhandled_branch` and `Invalid_condition_unhandled_branch_and_unhandled_condition`

**Comment by DeanArbel** [comment]
@Itay4 ran the validation on all content - didn't encounter an error with the playbooks.

**Comment by Itay4** [comment]
@DeanArbel see my comment https://github.com/demisto/demisto-sdk/pull/239#issuecomment-596225551
the first error message is not clear enough, we can improve it
what is `default`? what is the issue in this case?

**Comment by DeanArbel** [comment]
> @DeanArbel see my comment [#239 (comment)](https://github.com/demisto/demisto-sdk/pull/239#issuecomment-596225551)
> the first error message is not clear enough, we can improve it
> what is `default`? what is the issue in this case?

1. Changed it, please see screenshot.
2. `#default#` is the `else` condition. I removed the # chars, and replaced default with else. Please see screenshot.

image

**Pull request merged**